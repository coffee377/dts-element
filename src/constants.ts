import * as ts from 'typescript';
import {create_native_type, INativeType} from './types/native-type';

export enum ElementKind {
  GenericDeclaration,
  GenericType,
  NativeType,
  ArrayType,
  Parameter,
}

export const any_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)});
export const never_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword)});

export const null_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword)});
export const undefined_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.UndefinedKeyword)});
export const void_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword)});

export const boolean_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword)});
export const number_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)});
export const string_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)});
export const symbol_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.SymbolKeyword)});
export const object_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword)});

export const this_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ThisKeyword)});
