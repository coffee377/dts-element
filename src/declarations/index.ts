export * from './class-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './interface-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
