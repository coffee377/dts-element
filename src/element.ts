export abstract class Element<Parameters extends {}> {

  public parameters: Parameters;

  constructor(parameters: Parameters) {
    this.parameters = parameters;
  }

  public get<Key extends keyof Parameters>(key: Key): Parameters[Key] {
    return this.parameters[key];
  }

  public set(values: Partial<Parameters>): this {
    Object.keys(values).forEach((key: keyof Parameters) => {
      const value = values[key];
      if (value !== undefined) {
        this.parameters[key] = value;
      }
    });
    return this;
  }

  public clone(is_deep_clone: boolean = false): this {
    const element = Object.create(this.constructor.prototype);
    element.parameters = {};
    element.copy(this, is_deep_clone);
    return element;
  }

  public copy(element: this, is_deep_copy: boolean = false): this {
    this.parameters = this._clone_parameters(element.parameters, is_deep_copy);
    return this;
  }

  private _clone_parameters(parameters: Parameters, is_deep_clone: boolean): Parameters {
    return Object.keys(parameters).reduce((cloned_parameters: Partial<Parameters>, key: keyof Parameters) => {
      const value = parameters[key];
      const cloned_value = this._clone_value(value, is_deep_clone);
      cloned_parameters[key] = cloned_value;
      return cloned_parameters;
    }, {} as any);
  }

  private _clone_value<T>(value: T, is_deep_clone: boolean): T {
    return (typeof value !== 'object')
      ? value
      : (value instanceof Element)
        ? value.clone(is_deep_clone)
        : (value instanceof Array)
          ? value.map(<U>(x: U): U => this._clone_value(x, is_deep_clone)) as any
          : value; // TODO
  }

}
