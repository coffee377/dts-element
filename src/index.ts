export * from './collections';
export * from './constants';
export * from './declarations/index';
export * from './element';
export * from './emit';
export * from './transform';
export * from './types/index';
