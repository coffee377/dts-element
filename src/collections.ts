import * as declarations from './declarations/index';
import * as types from './types/index';

export type IDeclaration =
  | declarations.IFunctionDeclaration
  | declarations.IGenericDeclaration
  | declarations.IParameterDeclaration
  | declarations.IVariableDeclaration
;

export type IType =
  | types.IArrayType
  | types.IConstructorType
  | types.IFunctionType
  | types.IGenericType
  | types.IIntersectionType
  | types.IKeyofType
  | types.ILiteralType
  | types.INativeType
  | types.ITypeofType
  | types.ITupleType
  | types.IUnionType
;
