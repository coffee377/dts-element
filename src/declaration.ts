import {Element} from './element';
import {line_map} from './helpers/line-map';

export interface IDeclarationRequiredParameters {
  name: string;
}

export interface IDeclarationOptionalParameters {
  jsdoc: string;
}

export type AnyDeclaration = Declaration<any, any>;

export abstract class Declaration<RequiredParameters extends {}, OptionalParameters extends {}> extends Element
    <IDeclarationRequiredParameters & RequiredParameters, IDeclarationOptionalParameters & OptionalParameters> {

  public get default_declaration_parameters(): IDeclarationOptionalParameters {
    return {
      jsdoc: '',
    };
  }

  public emit_jsdoc(): string {
    const {jsdoc} = this.parameters;
    const content = line_map(jsdoc, (line: string) => ` * ${line}`);
    return (content.length === 0)
      ? content
      : `/**\n${content}\n */\n`;
  }

}
