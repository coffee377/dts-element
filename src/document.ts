import {Element} from './element';
import {RootElement} from './sets';
import {trim_every_line} from './helpers/trim-every-line';
import {AnyElement} from './element';

export interface IDocumentRequiredParameters {
  elements: RootElement[];
}

// tslint:disable-next-line no-empty-interface
export interface IDocumentOptionalParameters {}

export class Document extends Element<IDocumentRequiredParameters, IDocumentOptionalParameters> {

  public get default_parameters(): IDocumentOptionalParameters {
    return {};
  }

  public _emit(_container: AnyElement): string {
    const {elements} = this.parameters;
    const content = elements.map((element: RootElement) => element.emit()).join('\n');
    return trim_every_line(content);
  }

}
