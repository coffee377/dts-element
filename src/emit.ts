import * as ts from 'typescript';
import {ElementKind} from './constants';
import {IElement} from './element';
import {emit_triple_slash_reference, ITripleSlashReference} from './others/triple-slash-reference';
import {transform} from './transform';

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) => {
  switch (element.kind) {
    case ElementKind.TripleSlashReference:
      return emit_triple_slash_reference(element as ITripleSlashReference);
    default:
      const ts_node = transform(element);
      const source_file = ts.createSourceFile('', '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
      const printer = ts.createPrinter(options);
      return printer.printNode(ts.EmitHint.Unspecified, ts_node, source_file);
  }
};
