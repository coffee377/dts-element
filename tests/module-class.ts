jest.disableAutomock();

import * as dts from '../src/index';

// template: https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-class-d-ts.html

const interface_MyClassMethodOptions = new dts.InterfaceDeclaration({
  name: 'MyClassMethodOptions',
  members: [
    new dts.ObjectMember({
      owned: new dts.VariableDeclaration({
        name: 'width',
        type: dts.number_type,
        optional: true,
      }),
    }),
    new dts.ObjectMember({
      owned: new dts.VariableDeclaration({
        name: 'height',
        type: dts.number_type,
        optional: true,
      }),
    }),
  ],
});

const namespace_MyClass = new dts.NamespaceDeclaration({
  name: 'MyClass',
  jsdoc: 'If you want to expose types from your module as well, you can\nplace them in this block.',
  children: [interface_MyClassMethodOptions],
});

const class_MyClass = new dts.ClassDeclaration({
  name: 'MyClass',
  jsdoc: 'Write your module\'s methods and properties in this class',
  members: [
    new dts.ClassMember({
      owned: new dts.Constructor({
        parameters: [
          new dts.Parameter({name: 'someParam', type: dts.string_type, flag: dts.ParameterFlags.OPTIONAL}),
        ],
      }),
    }),
    new dts.ClassMember({
      owned: new dts.VariableDeclaration({
        name: 'someProperty',
        type: new dts.ArrayType({owned: dts.string_type}),
      }),
    }),
    new dts.ClassMember({
      owned: new dts.FunctionDeclaration({
        name: 'myMethod',
        parameters: [
          new dts.Parameter({
            name: 'opts',
            type: new dts.SubType({
              namespaces: [namespace_MyClass],
              target: new dts.InterfaceType({owned: interface_MyClassMethodOptions}),
            }),
          }),
        ],
        return: dts.number_type,
      }),
    }),
  ],
});

const export_as = new dts.ExportAsNamespace({name: 'myClassLib'});
const export_equal = new dts.ExportEqual({value: class_MyClass});

const document = new dts.Document({children: [
  export_as,
  export_equal,
  class_MyClass,
  namespace_MyClass,
]});

it('should return correctly', () => {
  expect(document.emit()).toMatchSnapshot();
});
