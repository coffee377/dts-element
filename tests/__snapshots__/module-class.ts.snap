// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
"export as namespace myClassLib;
export = MyClass;
/**
 * Write your module's methods and properties in this class
 */
class MyClass {
  constructor(someParam?: string);
  someProperty: string[];
  myMethod(opts: MyClass.MyClassMethodOptions): number;
}
/**
 * If you want to expose types from your module as well, you can
 * place them in this block.
 */
declare namespace MyClass {
  interface MyClassMethodOptions {
    width?: number;
    height?: number;
  }
}"
`;
